### PR TITLE
2015.8 fix pkg latest

### PR DESCRIPTION
**Pull request by joejulian**
### What does this PR do?

Fix pkg.latest idempotency

When installing multiple packages, pkg.latest was failing on arch and centos if the packages were already installed. This inverts the logic for determining if a package needs updated to make it simpler.

This uncovered a flaw in the pkg.latest_version module for yum where it was not returning versions if a package was installed. This fixes that as well.
### What issues does this PR fix or reference?

Issue # 35747
### Previous Behavior

Subsequent calls to pkg.latest with multiple packages would fail on arch and centos if the pkgs were successfully installed the first time.
### New Behavior

Subsequent calls to all pkg.\* state functions are now idempotent.
### Tests written?

Yes

### CODE REVIEW

**Comment by cachedout** [comment]
@terminalmage and @lorengordon You've both been in this code lately. Any thoughts?

File: salt/modules/yumpkg.py [hunk 1]
```diff
@@ -580,8 +580,9 @@ def latest_version(*names, **kwargs):
     cmd = [_yum(), '--quiet']
     cmd.extend(repo_arg)
     cmd.extend(exclude_arg)
-    cmd.extend(['list', 'available'])
+    cmd.extend(['list'])
     cmd.extend(names)
+    log.debug(cmd)
```
Comment (lorengordon) [inline salt/modules/yumpkg.py]: Was log statement left here intentionally? If so, might be nice to put a little more context around the message. Though, I thought the cmd module would print the command anyway with debug logging?

**Comment by lorengordon** [comment]
@cachedout I had eyeballed this one earlier to see if it was related to anything I might have worked on, but it's not; I was only working in the win_pkg.py module. I'm afraid I'm not familiar enough with this section of code to review it in much depth without putting in the time to study it more thoroughly...

File: salt/modules/yumpkg.py (hunk 1 above)
Comment (joejulian) [inline salt/modules/yumpkg.py]: No, it wasn't. I missed that. Thanks.

**Comment by joejulian** [comment]
To explain the logic change:
desired_pkgs = ['foo','bar']
avail = {'foo': '1.0.0-1', 'bar': '2.0.0-1'}
cur = {'foo': '1.0.0-1', 'bar': '2.0.0-1'}

So originally we did an if not avail[pkg]. pkg *is* avail, so we skip right down to else and add avail[pkg] to targets[pkg]. But we didn't want it as a target because it was already installed.

This positive logic makes choices based on what things *are* instead of what they're *not*, which is much easier to follow.

If the package is available (avail[pkg]) then if the package is not installed (not cur[pkg]) or if the version is newer, add it to the targets.

Otherwise, since the package is not available, error.

Removing the 'available' from the package listing is because we actually need a list of all the available packages, even if they're installed. Yum would fail to return a package name and version if it was installed already.

As for the tests, I've really added nothing. The same pre-existing tests are just executed twice to ensure they still succeed the second time.

File: salt/modules/yumpkg.py [hunk 2]
```diff
@@ -580,7 +580,7 @@ def latest_version(*names, **kwargs):
     cmd = [_yum(), '--quiet']
     cmd.extend(repo_arg)
     cmd.extend(exclude_arg)
-    cmd.extend(['list', 'available'])
+    cmd.extend(['list'])
```
Comment (terminalmage) [inline salt/modules/yumpkg.py]: According to the yum manpage:

```
       yum list [all | glob_exp1] [glob_exp2] [...]
              List all available and installed packages.

       yum list available [glob_exp1] [...]
              List  all  packages  in  the  yum  repositories  available  to   be
              installed.
```

Given that the [stated behavior of the pkg.latest_version function](https://docs.saltstack.com/en/latest/ref/modules/all/salt.modules.yumpkg.html#salt.modules.yumpkg.latest_version) is not to return a version when the package is already installed, the code is already working as intended. This change would in fact _break_ the ``pkg.latest`` state for platforms using yum/dnf, and as such, I can't recommend merging this pull request until it is removed.

**Comment by terminalmage** [comment]
Running the tests a second time does nothing to confirm idempotence, as ``assertSaltTrueReturn`` simply checks the ``result`` value of the state, which is ``True`` even when there are changes. So, I really don't see _any_ benefit to running the states a second time.

A better way of confirming idempotence would be to run a ``pkg.version`` on the targeted package after the state is run, and do a ``self.assertEqual()`` to ensure that it matches the version we expected to be installed.

**Comment by joejulian** [comment]
"Does nothing" isn't entirely accurate. It does find the bug that I intended to fix, but you're right. It could be better. The bug that I was looking for was one where pkg.latest multi would Fail on the second run for both Arch and Centos. This does find that.

Perhaps just checking for and empty change set without failure would be a better idempotence test. I'd prefer that enhancement to the test be its own issue. This fixes and tests the bug I reported.

**Comment by joejulian** [comment]
Ok, I don't care if it works anymore. I'll just do a for loop.

File: salt/states/pkg.py [hunk 3]
```diff
@@ -1624,18 +1625,18 @@ def latest(
 
     targets = {}
     problems = []
+    cmp_func = __salt__.get('pkg.version_cmp')
     for pkg in desired_pkgs:
```
Comment (terminalmage) [inline salt/states/pkg.py]: The changes below are based on an incorrect assumption of how ``pkg.latest_version`` is supposed to work, likely due to there being a bug in pacman.py which I have fixed in https://github.com/saltstack/salt/pull/35880.

There is no need to do a version comparison here, this was deliberately removed recently and should not be added back. If ``avail[pkg]`` evaluates as false, then there is either a) no upgrade available, or b) a package by that name does not exist.

If ``avail[pkg]`` evaluates as true, then there _is_ an update available and the package should be added to the installation targets. The mere fact that ``avail[pkg]`` contains something other than an empty string is enough to determine that the package needs to be updated, there is no need for a version comparison (which is why it was removed).

I added some clarifying comments to this block of code in https://github.com/saltstack/salt/pull/35882, as well as fixed the condition for Gentoo USE flags.

**Pull request closed without merging**